src/plugins/intel_cpu/src/transformations/snippets/x64/pass/lowered/external_repacking_adjuster.cpp

```diff
@@ -20,53 +20,129 @@ BrgemmExternalRepackingAdjuster::BrgemmExternalRepackingAdjuster(const ov::snipp
+        fit_into_L2 ? CPURuntimeConfig::RepackingImplType::IN_PARALLEL : CPURuntimeConfig::RepackingImplType::SEPARATE;
+
+    const auto is_impl_parallel = cpu_config->repacking_impl_type == CPURuntimeConfig::RepackingImplType::IN_PARALLEL;
+
+    for (const auto& p : m_executors) {
+        const auto& i = p.first;
         const auto& shape = cpu_config->io_shapes[i];
-        const auto& K = *++shape.rbegin();
-        const auto& N = *shape.rbegin();
-
-        const auto& precision = linear_ir.get_parameters()[i]->get_node()->get_output_element_type(0);
-        const auto vnni_factor = brgemm_utils::compute_vnni_factor(precision);
-        const size_t brgemm_kernel_rank = 2;
-        // Firstly, batch dims are set
-        VectorDims requested_blocked_shape(shape.begin(), shape.end() - brgemm_kernel_rank);
-        // Then, the blocked dims are formed
-        requested_blocked_shape.insert(requested_blocked_shape.end(),
-                                       {snippets::utils::div_up(K, vnni_factor),
-                                        std::max(N, brgemm_utils::repacking::compute_inner_n_block(precision)),
-                                        vnni_factor});
-
-        VectorDims requested_order(shape.size() - brgemm_kernel_rank);
-        std::iota(requested_order.begin(), requested_order.end(), 0);
-        const auto last_idx = shape.size() - 1;
-        requested_order.insert(requested_order.end(), {last_idx - 1, last_idx, last_idx - 1});
-
-        optimal_descs[i] =
-            std::make_shared<CpuBlockedMemoryDesc>(precision, Shape(shape), requested_blocked_shape, requested_order);
-
-        ov::snippets::VectorDims shape_for_offset(cpu_config->tensor_rank - shape.size(), 1);
-        shape_for_offset.insert(shape_for_offset.end(), requested_blocked_shape.begin(), requested_blocked_shape.end());
-        m_configurator->compute_offsets(shape_for_offset, i, 0);
+        auto& repacked_in = cpu_config->repacked_inputs[i];
+
+        const auto& prc = linear_ir.get_parameters()[i]->get_node()->get_output_element_type(0);
+        auto planar_shape = ov::snippets::utils::get_planar_vdims(shape, cpu_config->io_layouts[i]);
+        auto blk_shape = get_blk_shape(planar_shape, prc);
+        // In parallel impl, each thread needs buffer with only shape [K_blk, N_blk, VNNI] to store repacking data
+        if (is_impl_parallel) {
+            std::fill(planar_shape.rbegin() + brgemm_kernel_rank, planar_shape.rend(), 1);
+            std::fill(blk_shape.rbegin() + brgemm_kernel_rank + 1, blk_shape.rend(), 1);
+        }
+        const auto order = get_blk_order(planar_shape.size());
+        const auto desc = std::make_shared<CpuBlockedMemoryDesc>(prc, Shape(planar_shape), blk_shape, order);
+
+        // Save original input offsets for input before repacking.
+        const auto in_offsets =
+            shape == cpu_config->latest_shapes[i] ? repacked_in.in_offsets() : cpu_config->io_data_offsets[i];
```

Comment:
Why can't we always use `cpu_config->io_data_offsets[i]`? If shape is equal to latest shape, `repacked_in.in_offsets()` and `cpu_config->io_data_offsets[i]` should be equal, right?